File: services/omai-control-plane/internal/domain/mcp_test.go
package domain

import "testing"

func TestMCPServerValidation(t *testing.T) {
	valid := []MCPServer{
		{ID: "local", WorkspaceID: "workspace", Name: "Local", Transport: "stdio", Command: "mcp-local"},
		{ID: "remote", WorkspaceID: "workspace", Name: "Remote", Transport: "streamable-http", URL: "https://example.com/mcp"},
	}
	for _, server := range valid {
		if err := server.Validate(); err != nil {
			t.Fatalf("valid server %#v: %v", server, err)
		}
	}
	invalid := []MCPServer{
		{ID: "local", WorkspaceID: "workspace", Name: "Local", Transport: "stdio"},
		{ID: "remote", WorkspaceID: "workspace", Name: "Remote", Transport: "sse", URL: "file:///secret"},
		{ID: "remote", WorkspaceID: "workspace", Name: "Remote", Transport: "sse", URL: "https://user@example.com/mcp"},
		{ID: "bad\nname", WorkspaceID: "workspace", Name: "Bad", Transport: "stdio", Command: "mcp"},
	}
	for _, server := range invalid {
		if err := server.Validate(); err == nil {
			t.Fatalf("invalid server was accepted: %#v", server)
		}
	}
}

File: services/omai-control-plane/sdk/typescript/test/mcp.test.ts
import { createRouterTransport } from "@connectrpc/connect"
import { describe, expect, it } from "vitest"
import { createOMAIClientFromTransport } from "../src/client.js"
import { MCPService, type MCPServerInfo } from "../src/gen/uab/v1/native_pb.js"

describe("OMAI MCP facade", () => {
  it("owns durable configuration without claiming runtime connectivity", async () => {
    const stored = new Map<string, MCPServerInfo>()
    const transport = createRouterTransport((router) => {
      router.service(MCPService, {
        list: () => ({ servers: [...stored.values()] }),
        upsert: (request) => {
          const server = { ...request.server!, workspaceId: request.workspaceId }
          stored.set(server.id, server)
          return { server }
        },
        delete: (request) => ({ deleted: stored.delete(request.serverId) }),
      })
    })
    const mcp = createOMAIClientFromTransport(transport).mcp

    await expect(
      mcp.upsert("workspace-1", {
        id: "docs",
        name: "Docs",
        transport: "stdio",
        command: "mcp-docs",
        enabled: false,
      }),
    ).resolves.toMatchObject({ id: "docs", workspaceId: "workspace-1", enabled: false })
    await expect(mcp.list("workspace-1")).resolves.toHaveLength(1)
    await expect(mcp.delete("workspace-1", "docs")).resolves.toBe(true)
    await expect(mcp.delete("workspace-1", "docs")).resolves.toBe(false)
  })

  it("rejects incomplete commands and credential-bearing URLs", async () => {
    const transport = createRouterTransport((router) => router.service(MCPService, {}))
    const mcp = createOMAIClientFromTransport(transport).mcp

    await expect(mcp.upsert("workspace-1", { id: "bad", name: "Bad", transport: "stdio" })).rejects.toThrow(TypeError)
    await expect(
      mcp.upsert("workspace-1", {
        id: "remote",
        name: "Remote",
        transport: "streamable-http",
        url: "https://secret@example.com/mcp",
      }),
    ).rejects.toThrow(TypeError)
  })
})
